src/planning/planning/utils/path_3d_rotation.py

```diff
@@ -1,55 +1,197 @@
+import matplotlib
+matplotlib.use('Agg')
+
 import numpy as np
 from scipy.interpolate import make_interp_spline
-from scipy.spatial.transform import Rotation, RotationSpline
+from scipy.spatial.transform import Rotation
+from scipy.optimize import root_scalar
+
 import matplotlib.pyplot as plt
 from matplotlib.animation import FuncAnimation
-from mpl_toolkits.mplot3d import Axes3D
 
-# Example points (x, y)
+import os
+
+if not os.path.exists("./out"):
+    os.mkdir("./out") # Make the output folder to store the visualization of path generation
+
+# Find the maximum value of a B-spline between t_start and t_end.
+def find_maximum_bspl(spline, t_start, t_end, num_points=1000):
+    t_values = np.linspace(t_start, t_end, num_points)
+    spline_values = spline(t_values)
+    abs_spline_values = np.abs(spline_values)
+    max_value_index = np.argmax(abs_spline_values)
+    max_abs_value = abs_spline_values[max_value_index]
+    max_time = t_values[max_value_index]
+
+    derivative_spline = spline.derivative()
+
+    def derivative_root_finder(t):
+        return derivative_spline(t)
+
+    critical_points = []
+    for i in range(num_points - 1):
+        if t_values[i] < t_start or t_values[i + 1] > t_end:
+            continue
+        try:
+            root_result = root_scalar(
+                derivative_root_finder, bracket=[t_values[i], t_values[i + 1]], method='brentq'
+            )
+            if root_result.converged and t_start <= root_result.root <= t_end:
+                critical_points.append(root_result.root)
+        except ValueError:
+            pass
+
+    evaluation_points = [t_start, t_end] + critical_points
+    for t in evaluation_points:
+        spline_value = spline(t)
+        abs_spline_value = np.abs(spline_value)
+        if abs_spline_value > max_abs_value:
+            max_abs_value = abs_spline_value
+            max_time = t
+
+    return max_abs_value, max_time
+
+def make_interp_spline_with_constraints(x, y, k=3, t=None, bc_type=None, v_max=None, a_max=None):
+    spline_pos = make_interp_spline(x=x, y=y, k=k, t=t, bc_type=bc_type)
+
+    if v_max is None or a_max is None:
+        return spline_pos, x
+
+    spline_vel = spline_pos.derivative()
+    spline_acc = spline_vel.derivative()
+    factors = []
+    for i in range(len(x) - 1):
+        factors.append(max(find_maximum_bspl(spline_vel, x[i], x[i+1])[0]/v_max, (find_maximum_bspl(spline_acc, x[i], x[i+1])[0]/a_max)**0.5))
+    
+    x_new = [x[0]]
+    for i in range(len(factors)):
+        x_new.append(x_new[-1] + (x[i + 1] - x[i]) * factors[i])
+
+    return make_interp_spline(x=x, y=y, k=k, t=t, bc_type=bc_type), x_new
+
+# 5 example points (x, y, z, theta_x, theta_y, theta_z)
```

Comment:
Make sure to nest these tests with if __name__=="__main__" so that they don't run when the file is imported.